### ISSUE TITLE
Question: Participants defined in dynamically loaded lib can't be discovered [9129]

### ISSUE THREAD

**Issue by alireza-hosseini** [eProsima/Fast-DDS#1343]
Please assume we have a shared library which internally defines a participant, subscriber, and publisher.
This shared library will be dynamically loaded during runtime. Now we have a google test which loads this shared library
and tries to test its publish and subscribe functionality using external publisher/subscribers defined in the test itself.
something like below pseudo-code:

```cpp
int domain = 0;
std::string topic = "status";
StatusMessage foobar = generateTestMsg();

//This library defines its own participant inside
auto sharedLibA = loadLib(domain, topic, nullptr);
auto sharedLibB = loadLib(domain, topic, []() 
{
   EXPECT_EQ(msg, foobar);
});

auto participant = createParticipant(domain);
auto publisherA = createPublisher(participant, topic);
auto subscriberA = createSubscriber(participant, topic, [](StatusMessage *msg)
{
   EXPECT_EQ(msg, foobar):
});

//send foobar using loaded lib and expect external `subscriberA` would receive it
sharedLibA.sendMsg(foobar);

//send foobar using external publisher and expect lib's subscriber receive it
publisherA.sendMsg(foobar);
```

The problem here is that the loaded libraries do not discover externally defined publishers and subscribers (and even the participant because I defined a `ParticipantListener` and no event triggered!) and also the other way around. In addition when I query the number of available publishers or subscribers from the external publisher and subscriber it's always one (just the externals are counted while there should actually be three publishers and three subscribers!),
So to summarize, It means I can't communicate send/receive any message to/from dynamically loaded libraries.
I'm sure that I'm using the same domain and even the same attributes for all the fastdds entities here, so I suspect there might be some tricks that I need to know about using fastdds in a shared library. 
So any help or hints would be really appreciated.

**Comment by IkerLuengo** [comment]
Hi @alireza-hosseini,

There should be no special actions needed to communicate participants in and out of a shared library. In any case, I'll try to build a test to reproduce your use case, and get back to you.

**Comment by IkerLuengo** [comment]
Hi @alireza-hosseini,

I have not been able to reproduce your problem. I can correctly discover, match and send and receive messages from participants defined inside the DLL and participants defined outside the DLL.

As a reference, here is the layout of the test I performed:

DLL exposes a method `DLLInit`. This method:
  - Creates a `dll_publisher_participant` that holds a publisher for topic `MyTopic`.
  - Sets the publisher to publish a message every 500 ms
  - Creates a `dll_subscriber_participant` that holds a subscriber for the same topic `MyTopic`.
  - Publisher and subscriber have listeners that print a log every time they match with another peer.
  - Subscriber also prints a message every time it receives a message.

The application's `main` does the same (i.e., create a publisher and a subscriber for topic `MyTopic`, each on a separate participant, and set the publisher to publish every 500 ms)

All participants use the domain ID `0`, and no special configuration has been done .

On the logs I can see:
  - The publisher inside the DLL matches with both subscribers
  - The publisher in the `main` matches with both subscribers
  - The subscriber inside the DLL matches with both publishers and receives messages from both

In case it helps, I attach a tar file with the implemented test. It's basically the HelloWorld example distributed with Fast-DDS, modified according to the description above. It will probably compile if you uncompress it on `examples/C++`. It is not an elegant code and it does crash during the cleanup, but it is good enough to check your issue.

If you are still facing problems, please, reopen the issue and provide more information about how to reproduce it. Specifically, describe how your use case differs from the attached implementation.

[HelloWorldDll.tar.gz](https://github.com/eProsima/Fast-DDS/files/5155426/HelloWorldDll.tar.gz)
